File: Problems/0109. (M) Convert Sorted List to Binary Search Tree/Java/solution.java
class Solution {
  // recursive solution with two pointers
  public TreeNode sortedListToBST(ListNode head) {
    if (head == null)
      return null;
    if (head.next == null)
      return new TreeNode(head.val);
    return recurse(head, null);
  }

  private TreeNode recurse(ListNode head, ListNode tail) {
    if (head == tail)
      return null;
    
    ListNode slow = head;
    ListNode fast = head;

    while (fast != tail && fast.next != tail) {
      slow = slow.next;
      fast = fast.next.next;
    }

    // recursively construct subtrees
    TreeNode left = recurse(head, slow);
    TreeNode right = recurse(slow.next, tail);
    return new TreeNode(slow.val, left, right);
  }
}

File: Problems/0109. (M) Convert Sorted List to Binary Search Tree/Python3/solution.py
class Solution:
  # recursive solution with two pointers
  def sortedListToBST(self, head):
    # sanity checks
    if not head:
      return None
    if not head.next:
      return TreeNode(head.val)

    return self.recurse(head, None)
    
  def recurse(self, head, tail):
    if head is tail:
      return None
    slow = fast = head

    while fast is not tail and fast.next is not tail:
      slow = slow.next
      fast = fast.next.next
    
    # recursively construct subtrees
    left = self.recurse(head, slow)
    right = self.recurse(slow.next, tail)
    return TreeNode(slow.val, left, right)

